Serve in container docs (#15814)
PrefectHQ/prefect: docs/3.0/deploy/static-infrastructure-examples/docker.mdx

# Set the command to run your application
CMD ["python", "serve_retrieve_github_stars.py"]
```

```dockerfile Using uv
# Use the official Python image with uv pre-installed
FROM ghcr.io/astral-sh/uv:python3.12-bookworm-slim

# Set the working directory
WORKDIR /app

# Set environment variables
ENV UV_SYSTEM_PYTHON=1
ENV PATH="/root/.local/bin:$PATH"

# Copy only the requirements file first to leverage Docker cache
COPY requirements.txt .

# Install dependencies
RUN --mount=type=cache,target=/root/.cache/uv \
    uv pip install -r requirements.txt

# Copy the rest of the application code
COPY serve_retrieve_github_stars.py .

# Set the entrypoint
ENTRYPOINT ["python", "serve_retrieve_github_stars.py"]
```

</CodeGroup>

<Note>
Using `pip`, the image is built in about 20 seconds, and using `uv`, the image is built in about 3 seconds.

You can learn more about using `uv` in the [Astral documentation](https://docs.astral.sh/uv/guides/integration/docker/).
</Note>

## Building and running the container

Now that we have a flow and a Dockerfile, we can build the the image from the Dockerfile and run a container from this image.

### Build (and push) the image

> We can build the image using the `docker build` command, using the `-t` flag to specify a name for the image.

Reviewer: ```suggestion
We can build the image with the `docker build` command and the `-t` flag to specify a name for the image.
```

```bash
docker build -t my-flow-image .
```

> At this point, you may also want to push the image to a container registry like [Docker Hub](https://hub.docker.com/) or [GitHub Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry). Please see the registries' respective documentation for details on authentication and registry naming conventions.

Reviewer: ```suggestion
At this point, you may also want to push the image to a container registry such as [Docker Hub](https://hub.docker.com/) or [GitHub Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry). Please see the registries' respective documentation for details on authentication and registry naming conventions.
```

### Run the container

> In practice, you'll likely want to inject some environment variables into your container, so let's define a `.env` file:

Reviewer: ```suggestion
You'll likely want to inject some environment variables into your container, so let's define a `.env` file:
```

```bash .env
PREFECT_API_URL=<YOUR-API-URL>
PREFECT_API_KEY=<YOUR-API-KEY-IF-USING-PREFECT-CLOUD>
```

> Then, run the container in [detached mode](https://docs.docker.com/engine/reference/commandline/run/#detached-d) (i.e. in the background):

Reviewer: ```suggestion
Then, run the container in [detached mode](https://docs.docker.com/engine/reference/commandline/run/#detached-d) (in other words, in the background):
```

```bash
docker run -d --env-file .env my-flow-image
```

#### Verify the container is running

```bash
docker ps | grep my-flow-image
```

You should see your container in the list of running containers, note the `CONTAINER ID` as we'll need it to view logs.

#### View logs
```bash
docker logs <CONTAINER-ID>
```

You should see logs from your newly served process, with the link to your deployment in the UI.

### Stop the container

```bash
docker stop <CONTAINER-ID>
```

## Next steps

Congratulations! You have packaged and served a flow on a long-lived Docker container.

> You may now easily deploy this container to other infrastructures, like:

Reviewer: ```suggestion
You may now easily deploy this container to other infrastructures, such as:
```

---
title: Serve flows in a long-lived Docker container
description: Learn how to serve a flow in a long-lived Docker container
---

> The `.serve` method allows you to easily elevate a flow to a deployment, immediately executing scheduled work [as a local process](/3.0/deploy/run-flows-in-local-processes).

Reviewer 2: 'immediately executing scheduled work' seems like a bit of an oxymoron. I suggested an alternate wording, but feel free to do something else.
```suggestion
The `.serve` method allows you to schedule runs of your flow and execute those runs [in a local process](/3.0/deploy/run-flows-in-local-processes).
```

# Use the official Python image with uv pre-installed
FROM ghcr.io/astral-sh/uv:python3.12-bookworm-slim

# Set the working directory
WORKDIR /app

# Set environment variables
ENV UV_SYSTEM_PYTHON=1
ENV PATH="/root/.local/bin:$PATH"

# Copy only the requirements file first to leverage Docker cache
COPY requirements.txt .

# Install dependencies
RUN --mount=type=cache,target=/root/.cache/uv \
    uv pip install -r requirements.txt

# Copy the rest of the application code
COPY serve_retrieve_github_stars.py .

# Set the entrypoint
ENTRYPOINT ["python", "serve_retrieve_github_stars.py"]
```

</CodeGroup>

<Note>
Using `pip`, the image is built in about 20 seconds, and using `uv`, the image is built in about 3 seconds.

You can learn more about using `uv` in the [Astral documentation](https://docs.astral.sh/uv/guides/integration/docker/).
</Note>

## Build and run the container

Now that we have a flow and a Dockerfile, we can build the image from the Dockerfile and run a container from this image.

### Build (and push) the image

We can build the image with the `docker build` command and the `-t` flag to specify a name for the image.

```bash
docker build -t my-flow-image .
```

> At this point, you may also want to push the image to a container registry such as [Docker Hub](https://hub.docker.com/) or [GitHub Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry). Please see the registries' respective documentation for details on authentication and registry naming conventions.

Reviewer 2: ```suggestion
At this point, you may also want to push the image to a container registry such as [Docker Hub](https://hub.docker.com/) or [GitHub Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry). Please refer to each registry's respective documentation for details on authentication and registry naming conventions.
```

### Run the container

You'll likely want to inject some environment variables into your container, so let's define a `.env` file:

```bash .env
PREFECT_API_URL=<YOUR-API-URL>
PREFECT_API_KEY=<YOUR-API-KEY-IF-USING-PREFECT-CLOUD>
```

Then, run the container in [detached mode](https://docs.docker.com/engine/reference/commandline/run/#detached-d) (in other words, in the background):

```bash

> docker run -d --env-file .env my-flow-image

Reviewer 2: this is dope
